tests/model_forms/tests.py

```diff
@@ -585,6 +585,37 @@ class Meta:
         m2 = mf2.save(commit=False)
         self.assertEqual(m2.mode, '')
 
+    def test_default_not_populated_on_nonblank_field_in_cleaned_data(self):
+        class PubForm(forms.ModelForm):
+            mode = forms.CharField(max_length=255, required=False)
+
+            class Meta:
+                model = PublicationDefaults
+                fields = ('mode',)
+
+            def _set_clean_with_mode(self, mode_value):
+                def clean(self):
+                    self.cleaned_data['mode'] = mode_value
+                PubForm.clean = clean
```

Comment:
I would override `clean()` as described in the ticket, e.g.
```python
class PubForm(forms.ModelForm):
    mode = forms.CharField(max_length=255, required=False)
    mocked_mode = None

    def clean(self):
        self.cleaned_data['mode'] = self.mocked_mode
        return self.cleaned_data

    class Meta:
        model = PublicationDefaults
        fields = ('mode',)

default_mode = 'di'

pub_form = PubForm({})
pub_form.mocked_mode = 'de'
pub = pub_form.save(commit=False)
...
```

Reply:
This does overwrite `clean`. I wanted a programmatic way to do it, so that I didn't have to subclass `PubForm` 3 times to test 3 different values, e.g.

```py
class PubForm1(forms.ModelForm):
	mode = forms.CharField(max_length=255, required=False)

	class Meta:
		model = PublicationDefaults
		fields = ('mode',)

	def clean(self):
		self.cleaned_data['mode'] = 'de'
# ===

class PubForm2(forms.ModelForm):
	mode = forms.CharField(max_length=255, required=False)

	class Meta:
		model = PublicationDefaults
		fields = ('mode',)

	def clean(self):
		self.cleaned_data['mode'] = ''

# ===

class PubForm3(forms.ModelForm):
	mode = forms.CharField(max_length=255, required=False)

	class Meta:
		model = PublicationDefaults
		fields = ('mode',)

	def clean(self):
		self.cleaned_data['mode'] = None
```

Reply:
But yeah, it doesn't read well. Will fix.

Reply:
TBH this is not what I suggested, I wanted to use `mocked_mode` to keep code DRY. I pushed edits.